lib/rom/command.rb

```diff
@@ -0,0 +1,82 @@
+require 'rom/commands'
+
+module ROM
+  class Command < Commands::AbstractCommand
+    include Equalizer.new(:relation, :options)
+
+    extend DescendantsTracker
+    extend ClassMacros
+
+    defines :type, :relation, :repository, :input, :validator, :result, :register_as
+
+    repository :default
+    result :many
+
+    def self.build_class(name, relation, options = {}, &block)
+      type = options.fetch(:type) { name }
+      class_name = "ROM::Command[#{relation}][#{type}]"
+
+      ClassBuilder.new(name: class_name, parent: self).call do |klass|
+        klass.type(type)
+        klass.register_as(name)
+        klass.relation(relation)
+        klass.class_eval(&block) if block
+        options.each { |k, v| klass.send(k, v) }
+      end
+    end
+
+    def self.adapter(name = nil)
+      return @adapter unless name
+
+      unless type
+        raise ArgumentError, "#{self}.type must be specified before `adapter`"
+      end
+
+      commands = ROM.adapters.fetch(name).const_get(:Commands)
+      mod = command_mod(commands)
+
+      include(mod)
+
+      @adapter = name
+
+      self
+    end
+
+    def self.command_mod(commands)
+      case type
+      when :create then commands.const_get(:Create)
+      when :update then commands.const_get(:Update)
+      when :delete then commands.const_get(:Delete)
+      else
+        raise(
+          ArgumentError,
+          "#{type.inspect} is not a supported command type for #{inspect}"
+        )
+      end
```

Comment:
Could this be refactored into a `Hash` lookup, so that the duplication could be removed? The keys would be the symbols (`:create, :update, :delete`) and perhaps the values could be a be a proc representing the code in the `then` clause. If the key isn't found, the error can still be raised.